File: semgus-sketch/src/main/java/org/semgus/sketch/base/Param.java
package org.semgus.sketch.base;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.stream.Stream;

public record Param(@NotNull Id type, @NotNull Id id) implements Syntax {
  public static @NotNull Param param(@NotNull Id type, @NotNull Id id) {
    return new Param(type, id);
  }

  public static @NotNull Param plain(@NotNull Id type, @NotNull String id) {
    return param(type, Id.plain(id));
  }

  public static @NotNull Param intPlain(@NotNull String id) {
    return param(Id.plain("int"), Id.plain(id));
  }

  public static @NotNull Param bnd() {
    return Param.intPlain("bnd");
  }

  public static @NotNull List<Param> bnded0(@NotNull List<Param> params) {
    var bndParam = Param.intPlain("bnd");
    return Stream.concat(params.stream(), Stream.of(bndParam)).toList();
  }

  public static @NotNull Param bitPlain(@NotNull String id) {
    return param(Id.plain("bit"), Id.plain(id));
  }
}

File: semgus-sketch/src/main/java/org/semgus/sketch/syntax/Param.kt
package org.semgus.sketch.syntax

/**
 * Sketch parameter.
 */
internal data class Param(val type: Id, val id: Id) : Syntax()

internal fun param(type: Id, id: Id) = Param(type, id)
internal fun paramPlain(type: Id, name: String) = param(type, idPlain(name))
internal fun intPlain(name: String) = paramPlain(idPlain("int"), name)
internal fun bitPlain(name: String) = paramPlain(idPlain("bit"), name)

internal fun paramsBnded0(params: Sequence<Param>) = params + intPlain("bnd")
